### ISSUE TITLE
wxListCtrl can trigger a crash in wxQt

### ISSUE THREAD

**Issue by AliKet** [wxWidgets/wxWidgets#22998]
### Description

I don't quite understand but in:
https://github.com/wxWidgets/wxWidgets/blob/58f5a885ec20ea518f94e16d2fcff6d0b8decd17/src/qt/listctrl.cpp#L1089-L1093

calling `m_qtTreeWidget->setModel(nullptr);` end up calling this function:
https://github.com/wxWidgets/wxWidgets/blob/58f5a885ec20ea518f94e16d2fcff6d0b8decd17/src/qt/listctrl.cpp#L164
presumably due to updating the internal state of the Model/View? I'm not sure, the debugger shows this.

**The problem is** we are in the `wxListCtrl` destructor and the object cannot be used any longer as it is half destroyed. Nevertheless, `data()` uses this (half destroyed) object to get the imagelist via `GetImageList()` and, **BOOM**, a crash happens.

Putting this:
```
        if ( m_listCtrl->IsBeingDeleted() )
            return nullptr;
```
inside `GetImageList()` seems to solve the problem, but looking at `wxListCtrl` dtor one more time I conclude tha we can just delete m_model right away and don't call `m_qtTreeWidget->setModel(nullptr);` which is not necessary because the control is dying anyhow. moreover the view does not take ownership of the model see [setModel](https://doc.qt.io/qt-5/qabstractitemview.html#setModel) docs.

The crash can be seen by running `ListBookTestCase`  test or by applying this patch to minimal sample:
```
diff --git a/samples/minimal/minimal.cpp b/samples/minimal/minimal.cpp
index 501caf9096..4eb69c354a 100644
--- a/samples/minimal/minimal.cpp
+++ b/samples/minimal/minimal.cpp
@@ -26,6 +26,10 @@
     #include "wx/wx.h"
 #endif
 
+#include "wx/artprov.h"
+#include "wx/imaglist.h"
+#include "wx/listbook.h"
+
 // ----------------------------------------------------------------------------
 // resources
 // ----------------------------------------------------------------------------
@@ -64,6 +68,8 @@ public:
     void OnQuit(wxCommandEvent& event);
     void OnAbout(wxCommandEvent& event);
 
+    wxListbook* m_list;
+
 private:
     // any class wishing to process wxWidgets events must use this macro
     wxDECLARE_EVENT_TABLE();
@@ -170,6 +176,22 @@ MyFrame::MyFrame(const wxString& title)
     SetSizer(sizer);
 #endif // wxUSE_MENUBAR/!wxUSE_MENUBAR
 
+    m_list = new wxListbook(this, wxID_ANY, wxDefaultPosition, wxSize(400, 300));
+
+    wxSize size(32, 32);
+
+    auto list = new wxImageList(size.x, size.y);
+    list->Add(wxArtProvider::GetIcon(wxART_INFORMATION, wxART_OTHER, size));
+    list->Add(wxArtProvider::GetIcon(wxART_QUESTION, wxART_OTHER, size));
+
+    m_list->AssignImageList(list);
+
+    auto panel1 = new wxPanel(m_list);
+    auto panel2 = new wxPanel(m_list);
+
+    m_list->AddPage(panel1, "Panel 1", false, 0);
+    m_list->AddPage(panel2, "Panel 2", false, 1);
+
 #if wxUSE_STATUSBAR
     // create a status bar just for fun (by default with 1 pane only)
     CreateStatusBar(2);
@@ -188,6 +210,7 @@ void MyFrame::OnQuit(wxCommandEvent& WXUNUSED(event))
 
 void MyFrame::OnAbout(wxCommandEvent& WXUNUSED(event))
 {
+/*
     wxMessageBox(wxString::Format
                  (
                     "Welcome to %s!\n"
@@ -200,4 +223,12 @@ void MyFrame::OnAbout(wxCommandEvent& WXUNUSED(event))
                  "About wxWidgets minimal sample",
                  wxOK | wxICON_INFORMATION,
                  this);
+
+*/
+    delete m_list;
 }
```
In conclusion, I think we can just do this:
```
diff --git a/src/qt/listctrl.cpp b/src/qt/listctrl.cpp
index e4f91ba036..8c241c9651 100644
--- a/src/qt/listctrl.cpp
+++ b/src/qt/listctrl.cpp
@@ -1088,8 +1088,7 @@ void wxListCtrl::Init()
 
 wxListCtrl::~wxListCtrl()
 {
-    m_qtTreeWidget->setModel(nullptr);
-    m_model->deleteLater();
+    delete m_model;
 }
 
 bool wxListCtrl::SetForegroundColour(const wxColour& col)
```
which also solves the problem for me.

Thank you for looking at this.

### Platform and version information

- wxWidgets version  you use:
- wxWidgets port  you use:
- OS  and its version:

+ GTK version: 
+ Which GDK  backend is used:
+ Current theme:

**Comment by AliKet** [comment]
I forgot to mention that if `m_qtTreeWidget->setModel(nullptr);`   is really necessary:


than we can apply this patch which also works correctly:
```
diff --git a/src/qt/listctrl.cpp b/src/qt/listctrl.cpp
index e4f91ba036..669ef00ea9 100644
--- a/src/qt/listctrl.cpp
+++ b/src/qt/listctrl.cpp
@@ -1031,6 +1031,11 @@ void wxQtListTreeWidget::itemActivated(const QModelIndex &index)
     EmitListEvent(wxEVT_LIST_ITEM_ACTIVATED, index);
 }
 
+template<>
+void wxQtEventSignalHandler< QTreeView, wxListCtrl >::HandleDestroyedSignal()
+{
+    this->setModel(nullptr);
+}
 
 wxListCtrl::wxListCtrl()
 {
@@ -1088,8 +1093,7 @@ void wxListCtrl::Init()
 
 wxListCtrl::~wxListCtrl()
 {
-    m_qtTreeWidget->setModel(nullptr);
-    m_model->deleteLater();
+    delete m_model;
 }
 
 bool wxListCtrl::SetForegroundColour(const wxColour& col)
```

**Comment by vadz** [comment]
I don't know enough about Qt to know if the delayed deletion is necessary or not. We could try deleting it immediately and see if this results in any problems.

But I don't quite understand this part:

> The problem is we are in the wxListCtrl destructor and the object cannot be used any longer as it is half destroyed. Nevertheless, data() uses this (half destroyed) object to get the imagelist via GetImageList() and, BOOM, a crash happens.

First of all, `wxListCtrl` is being destroyed but hasn't been destroyed yet, so calling its virtual functions etc should be still safe. Second, how do we end up in `GetImageList()` anyhow, does the debugger show you this?

**Comment by AliKet** [comment]
> I don't know enough about Qt to know if the delayed deletion is necessary or not. We could try deleting it immediately and see if this results in any problems.
> 

Deleting the model immediately or later doesn't really matter, as both works without any problem. `m_qtTreeWidget->setModel(nullptr);` is the probelmatic line as shown by the debugger (see below), and I think removing that line doesn't seem to have any (bad) consequences!

> But I don't quite understand this part:
> 
> > The problem is we are in the wxListCtrl destructor and the object cannot be used any longer as it is half destroyed. Nevertheless, data() uses this (half destroyed) object to get the imagelist via GetImageList() and, BOOM, a crash happens.
> 
> First of all, `wxListCtrl` is being destroyed but hasn't been destroyed yet, so calling its virtual functions etc should be still safe. Second, how do we end up in `GetImageList()` anyhow, does the debugger show you this?

I'm intrigued too!
Anyway here is the relevant backtrace from running the minimal sample under the debugger (with the patch above applied):
```
Thread 1 "minimal" received signal SIGSEGV, Segmentation fault.
0xb797f7f6 in wxRefCounter::IncRef (this=0x34e0c5c6) at ../include/wx/object.h:235
235	    void IncRef() { m_count++; }
(gdb) bt
#0  0xb797f7f6 in wxRefCounter::IncRef() (this=0x34e0c5c6) at ../include/wx/object.h:235
#1  0xb797f8dd in wxObject::wxObject(wxObject const&) (this=0xbfffd0f4, other=...) at ../include/wx/object.h:387
#2  0xb7980cb8 in wxGDIObject::wxGDIObject(wxGDIObject const&) (this=0xbfffd0f4) at ../include/wx/gdiobj.h:41
#3  0xb79a14bc in wxBitmapBase::wxBitmapBase(wxBitmapBase const&) (this=0xbfffd0f4) at ../include/wx/bitmap.h:160
#4  0xb79a14f6 in wxBitmap::wxBitmap(wxBitmap const&) (this=0xbfffd0f4) at ../include/wx/qt/bitmap.h:15
#5  0xb79a0cb7 in wxGenericImageList::GetBitmap(int) const (this=0x9ac9b0, index=1) at ../src/generic/imaglist.cpp:176
#6  0xb79f674b in wxQtListModel::data(QModelIndex const&, int) const (this=0xa1cd50, index=..., role=1)
    at ../src/qt/listctrl.cpp:208
#7  0xb6d060a2 in QStyledItemDelegate::initStyleOption(QStyleOptionViewItem*, QModelIndex const&) const ()
    at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#8  0xb6d05590 in QStyledItemDelegate::sizeHint(QStyleOptionViewItem const&, QModelIndex const&) const ()
    at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#9  0xb6d4e77a in QTreeView::indexRowSizeHint(QModelIndex const&) const () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#10 0xb6d4febc in QTreeViewPrivate::itemHeight(int) const () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#11 0xb6d519e1 in QTreeViewPrivate::updateScrollBars() () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#12 0xb6d59e31 in QTreeView::updateGeometries() () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#13 0xb6d4a149 in QTreeView::columnCountChanged(int, int) () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#14 0xb6d55c26 in  () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#15 0xb60b2c6e in QMetaObject::activate(QObject*, int, int, void**) () at /usr/lib/i386-linux-gnu/sse2/libQt5Core.so.5
#16 0xb60b32dd in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) ()
    at /usr/lib/i386-linux-gnu/sse2/libQt5Core.so.5
#17 0xb6cec136 in QHeaderView::sectionCountChanged(int, int) () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#18 0xb6cf30ef in QHeaderView::initializeSections() () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#19 0xb6cf7282 in QHeaderView::reset() () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#20 0xb6cd90d8 in QAbstractItemView::setModel(QAbstractItemModel*) () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#21 0xb6cf7081 in QHeaderView::setModel(QAbstractItemModel*) () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#22 0xb6d57b19 in QTreeView::setModel(QAbstractItemModel*) () at /usr/lib/i386-linux-gnu/libQt5Widgets.so.5
#23 0xb79f122a in wxListCtrl::~wxListCtrl() (this=0x5c5450, __in_chrg=<optimized out>) at ../src/qt/listctrl.cpp:1103
#24 0xb7b5fdec in wxListView::~wxListView() (this=0x5c5450, __in_chrg=<optimized out>) at ../include/wx/listctrl.h:42
#25 0xb7b5fe1d in wxListView::~wxListView() (this=0x5c5450, __in_chrg=<optimized out>) at ../include/wx/listctrl.h:42
#26 0xb7bc16a6 in wxWindowBase::Destroy() (this=0x5c5450) at ../src/common/wincmn.cpp:572
#27 0xb7bc1874 in wxWindowBase::DestroyChildren() (this=0x4ad220) at ../src/common/wincmn.cpp:604
#28 0xb7a46a99 in wxWindow::~wxWindow() (this=0x4ad220, __in_chrg=<optimized out>) at ../src/qt/window.cpp:366
#29 0xb7aa0e00 in wxControlBase::~wxControlBase() (this=0x4ad220, __in_chrg=<optimized out>) at ../src/common/ctrlcmn.cpp:45
#30 0xb7994199 in wxControl::~wxControl() (this=0x4ad220, __in_chrg=<optimized out>) at ../include/wx/qt/control.h:11
#31 0xb7a0e9d7 in wxBookCtrlBase::~wxBookCtrlBase() (this=0x4ad220, __in_chrg=<optimized out>) at ../include/wx/bookctrl.h:55
#32 0xb7bdbf0e in wxNavigationEnabled<wxBookCtrlBase>::~wxNavigationEnabled() (this=0x4ad220, __in_chrg=<optimized out>)
    at ../include/wx/containr.h:189
#33 0xb7c0de87 in wxListbook::~wxListbook() (this=0x4ad220, __in_chrg=<optimized out>) at ../include/wx/listbook.h:39
#34 0xb7c0deb5 in wxListbook::~wxListbook() (this=0x4ad220, __in_chrg=<optimized out>) at ../include/wx/listbook.h:39
#35 0x0040a4aa in MyFrame::OnAbout(wxCommandEvent&) (this=0x574900) at ../../../samples/minimal/minimal.cpp:233
#36 0xb74f4c96 in wxAppConsoleBase::HandleEvent(wxEvtHandler*, void (wxEvtHandler::*)(wxEvent&), wxEvent&) const (this=0x44b1b0, handler=0x574900, func=
```
As you can see, lines **20**, **21** and **22** are due to calling `m_qtTreeWidget->setModel(nullptr);` inside `wxListCtrl::~wxListCtrl()`

**Comment by vadz** [comment]
Is the `model` (i.e. `this` pointer in e.g. frame 6) the same as `m_model` or not? If it is, I still don't understand why does it crash... If it isn't, I wonder what it is, then.

The problem with not calling `setModel(0)` is that if Qt ever calls any of model methods during the destruction, we'd be guaranteed to crash, so I'd still feel safer if we could do this. OTOH we already delete the old model immediately in `wxListCtrl::SetWindowStyleFlag()`, so we definitely should be able to do it here too instead of calling `deleteLater()`.

**Comment by AliKet** [comment]
> Is the `model` (i.e. `this` pointer in e.g. frame 6) the same as `m_model` or not? If it is, I still don't understand why does it crash... If it isn't, I wonder what it is, then.
> 

I checked it and indeed `this` pointer and `m_model` are the same.

> The problem with not calling `setModel(0)` is that if Qt ever calls any of model methods during the destruction, we'd be guaranteed to crash, so I'd still feel safer if we could do this. OTOH we already delete the old model immediately in `wxListCtrl::SetWindowStyleFlag()`, so we definitely should be able to do it here too instead of calling `deleteLater()`.

**Comment by vadz** [comment]
Something must be very wrong somewhere for `wxQtListModel::data()` to crash when it's called on the still perfectly valid pointer, but I just don't see what...

Anyhow, if you think it's ok to not call `setModel()`, let's not do it, it's definitely useless to refresh the control geometry when deleting it in any case.

**Comment by AliKet** [comment]
> Anyhow, if you think it's ok to not call `setModel()`, let's not do it, it's definitely useless to refresh the control geometry when deleting it in any case.

I'm not totally sure, I'll try one more time to figure out what's going on here before removing the questionable line.

Thank you for your time on this.

**Comment by AliKet** [comment]
Finally I found the culprit of this bug, it's the `wxListbook` and not `wxListCtrl` as I wrote in the title, sorry, which causes `wxQtListModel::data()` to crash.

**From the backtrace above:**
Inside `~wxListCtrl()` (line **23**) the image list (accessed by `data()`) is already freed after `wxBookCtrlBase` (which derives from `wxControl` and `wxWithImages`) and its base destructors run  (lines **31** and **32** - `wxWithImages` dtor not reported by gdb).

So `m_qtTreeWidget->setModel(nullptr);` is definitely innocent.

@vadz do you know how to properly handle this ? TIA!

**Comment by vadz** [comment]
Sorry, I still don't quite follow this... The call to `GetImageList()` in `wxQtListModel` ends up using `wxListCtrlBase::GetImageList()` which uses `wxListCtrlBase::GetImages()`, i.e. `m_imagesNormal` or `m_imagesSmall`, which are both members of `wxListCtrlBase` and haven't been destroyed yet. The `wxWithImages` in `wxBookCtrlBase` shouldn't be used here AFAICS -- is it really?

I don't have time to do it right now, but to debug this I'd build wxQt with ASAN because it should normally point out directly to where the already freed object is used and also, importantly, where it was freed.

**Comment by AliKet** [comment]
When creating the `listbook` it also creates a `listview`:
https://github.com/wxWidgets/wxWidgets/blob/58f5a885ec20ea518f94e16d2fcff6d0b8decd17/src/generic/listbkg.cpp#L83-L90

Then, when we create the `imagelist` and assign it to the `listbook` (using `AssignImageList()`), the `listbook` becomes the owner and will delete the `imagelist` when it get destroyed.

Now, when `wxListbook::OnImagesChanged()` is executed (as a result of calling `AssignImageList()`) it calls [list->SetImageList(GetImageList(), wxIMAGE_LIST_NORMAL);
](https://github.com/wxWidgets/wxWidgets/blob/58f5a885ec20ea518f94e16d2fcff6d0b8decd17/src/generic/listbkg.cpp#L286)
and this the problem: the `listview` will be left with a dangling pointer to an already freed `imagelist` when the `listbook` get destroyed.

The simplest and the immediate fix for this is to explicitly destroy the `m_bookctrl` created by the `listbook` ourselves. or, make `wxImageList` shareable and delete it only when the last reference reaches zero.

**Comment by vadz** [comment]
Thanks, I finally understand what's going on here. It's indeed a problem to leave `wxListCtrl` alive with a dangling pointer to an already destroyed `wxImageList` and the most logical solution would indeed be to just destroy `m_bookctrl` in `wxBookCtrlBase` dtor (which would need to be added). I don't think this can create any new problems because `m_bookctrl` is currently being destroyed only slightly later anyhow, when the base class destroys all its children, and it should fix this bug. Will you make a PR fixing this or should I do it myself? In any case, thanks a lot for finding and debugging this!

Just for completeness, `wxImageList` is already either non-owning (`Set`) or owning (`Assign`) and adding reference counting semantics on top of this would be confusing, without speaking of the fact that this class is deprecated (even if not formally so) and so I'd rather not add anything new to it.

**Comment by AliKet** [comment]
> Just for completeness, `wxImageList` is already either non-owning (`Set`) or owning (`Assign`) and adding reference counting semantics on top of this would be confusing, without speaking of the fact that this class is deprecated (even if not formally so) and so I'd rather not add anything new to it.

But how does a control using a `wxImageList` (which it doesn't own) know that it's using a valid image list and isn't being destroyed by its owner? because the widgets sample still crashes when the image list get destroyed here: https://github.com/wxWidgets/wxWidgets/blob/7d070a7e83acf7c2661cc6cdd8fa6bef63ff49e8/samples/widgets/notebook.cpp#L323

**Comment by vadz** [comment]
It doesn't know it, the owner of the image list is supposed to make sure that its lifetime is long enough. Which is indeed not the case here, so I guess we are going to need some Qt-specific workaround, after all :-(

The idea would be to ensure that the image list doesn't get used from the dtor, e.g. something like adding a check for `!m_listCtrl->IsBeingDeleted()` to `wxQtListModel::GetImageList()` and setting `m_isBeingDeleted` to true before resetting the model in the dtor. Ugly, but this should work, I think.

**Comment by AliKet** [comment]
> The idea would be to ensure that the image list doesn't get used from the dtor, e.g. something like adding a check for `!m_listCtrl->IsBeingDeleted()` to `wxQtListModel::GetImageList()` and setting `m_isBeingDeleted` to true before resetting the model in the dtor. Ugly, but this should work, I think.

With this patch:
```
diff --git a/src/qt/listctrl.cpp b/src/qt/listctrl.cpp
index e4f91ba036..878cd036fc 100644
--- a/src/qt/listctrl.cpp
+++ b/src/qt/listctrl.cpp
@@ -712,6 +712,16 @@ public:
 protected:
     wxImageList *GetImageList() const
     {
+        if ( m_listCtrl->IsBeingDeleted() )
+        {
+            // The image list (which we don't own) might have already been deleted
+            // at this point. e.g.: this can happen for a wxListbook (which uses
+            // a wxListView as controller buddy) because it deletes the image list
+            // when it's destroyed leaving the wxListView with invalid pointer to
+            // that image list.
+            return nullptr;
+        }
+
         const int requiredList = m_listCtrl->HasFlag(wxLC_SMALL_ICON)
             ? wxIMAGE_LIST_SMALL
             : wxIMAGE_LIST_NORMAL;
```
I can confirm that the crash disappears from the widgets sample (`listbook` page) as well as from `ListbookTestCase` test.

So this issue should be closed if you apply the patch to master.

Thank you once again for your time.

**Comment by AliKet** [comment]
@vadz sorry but there is also a problem with virtual `wxListCtrl` and also happens when destroying the control which results in a crash!

Virtual `wxListCtrl`s are supposed to override this function:
`    virtual wxString OnGetItemText(long item, long column) const;`
The base class version is made to fail if the derived class forgot to override it:
https://github.com/wxWidgets/wxWidgets/blob/48d226bb84efd8672de9ec8182a0dbc6b02bc90f/src/common/listctrlcmn.cpp#L252-L259

Now this line:
https://github.com/wxWidgets/wxWidgets/blob/48d226bb84efd8672de9ec8182a0dbc6b02bc90f/src/qt/listctrl.cpp#L862
inside `wxQtVirtualListModel::data()`  may end up calling the base class version and not the overriden when destroying the control because we are in fact inside the `wxListCtrl` destructor and this is really the nasty thing that can happen when **virtual functions are called from destructors**

I tried to `return` from data if `listCtrl->IsBeingDeleted()` is true but it doen't help unfortunately!
